arch: arm:  common Armv8-M support

ioannisg opened the pull request:
This PR includes the required changes in order to support
conditional compilation for Armv8-M architecture and its
supported Cortex cores, Cortex-M23 and Cortex-M33. Two
variants of the Armv8-M architecture are defined:
- the ARMV8_M_BASELINE (backwards compatible with Armv6-M),
- the ARMV8_M_MAINLINE (backwards compatible with Armv7-M).

In addition, the PR includes some minor style updates, based
on the updated ARM nomenclature for architecture definition.

**The PR could now be merged, once review is completed.** 

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -65,7 +65,7 @@ config ARMV6_M
 	select ATOMIC_OPERATIONS_C
 	select ISA_THUMB2
 	help
-	  This option signifies the use of an ARMv6-M processor implementation.
```
updated nomenclature from ARM

ioannisg commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -76,7 +76,7 @@ config ARMV7_M
 	select CPU_CORTEX_M_HAS_BASEPRI
 	select CPU_CORTEX_M_HAS_PROGRAMMABLE_FAULT_PRIOS
 	help
-	  This option signifies the use of an ARMv7-M processor implementation.
```
updated nomenclature from ARM

carlescufi commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -24,7 +24,7 @@
 #include <string.h>
 
-#ifdef CONFIG_ARMV6_M
+#if defined(CONFIG_ARMV6_M) || defined(CONFIG_ARMV8_M_BASELINE)
```
Maybe we could define some `ARMV6_V8_BASELINE` to avoid having to `||` all the time?

ioannisg replied:
@carlescufi  I do not have a strong opinion on that. Having the compile-time _OR_ directive is more descriptive and educative; a composite define saves space, and reduces risk of forgetting a directive.

galak replied:
I'm with @carlescufi - we should have a single option here, its less error prone.

How about CONFIG_ARMV6_AND_V8_BASELINE

ioannisg replied:
@carlescufi @galak  As a single Kconfig option, I would propose:

CONFIG_ARMV6_M_ARMV8_M_BASELINE (we should keep the _M_ in the name)

And, also, in Kconfig, I would not define this option directly, but as an option dependent on the "parent" options, as follows:


config ARMV6_M_ARMV8_M_BASELINE
    bool
    depends on ARMV6_M or ARMV8_M_BASELINE
    default n
    .........

Does this look good enough?

galak replied:
What do you mean by not define this option directly?

I think we should just 'select' it in the right place. depends on gets a little messy here.

ioannisg replied:
@galak I meant the following:
ARMV6_M, ARMV8_M_BASELINE would be "selected" by the corresponding Cortex cores (for example, CM0 and CM33, respectively). Then ARMV6_M_ARMV8_M_BASELINE would be "in"directly selected via the Kconfig option definition I presented above.

But perhaps, it is an overkill, and we could just select the common option in the cortex cores.

ioannisg replied:
@galak @carlescufi  I updated the PR based on your suggestion to use a composite Kconfig option.

carlescufi commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -42,7 +42,7 @@ void __weak relocate_vector_table(void)
 }
 
-#elif defined(CONFIG_ARMV7_M)
+#elif defined(CONFIG_ARMV7_M) || defined(CONFIG_ARMV8_M_MAINLINE)
```
Maybe we could define some `ARMV7_V8_MAINLINE` to avoid having to `||` all the time?

ioannisg replied:
@carlescufi  Same answer as above: I am open to that option.

carlescufi commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -65,7 +65,7 @@ config ARMV6_M
 	select ISA_THUMB2
 	help
-	  This option signifies the use of an ARMv6-M processor implementation.
+	  This option signifies the use of an Armv6-M processor implementation.
```
Has the case been changed for v6 and v7? Ithought the correct spelling were:
* ARMv6-M
* ARMv7-M
* Armv8-M

ioannisg replied:
Please, check https://developer.arm.com/products/architecture/m-profile. It seems that ARM has moved to the new style/nomenclature, regarding the Architecture namings.

galak replied:
While ARM has rebranded to Arm, I don't think it makes sense to make such changes in our code base.  I'm sure we are always going to have a mix, and I'm not keen on trying to match arm's marketing department on how to capitalize the letters in 'ARM'

ioannisg replied:
Makes sense. That is why I made the changes only in this specific file, because different spelling styles within the same file might cause some confusion. 

But we could revert these changes if you insist. No strong opinion on that.

galak replied:
lets revert such changes.

ioannisg replied:
Will do.

galak commented:
Where are the baseline and mainline terms defined by arm?

carlescufi commented:
@galak https://developer.arm.com/products/architecture/m-profile/docs/ddi0553/latest/armv8-m-architecture-reference-manual

galak reviewed (changes requested):
Can we split this into at least 2 comments.  One that deals with the ARMV6 -> BASELINE, ARMV7 -> MAINLINE stuff, than one that adds cortex-m23/m33.

carlescufi commented:
@galak s/comments/commits ?

galak commented:
> @galak https://developer.arm.com/products/architecture/m-profile/docs/ddi0553/latest/armv8-m-architecture-reference-manual
> 
> A PE with the Main Extension is also referred to as a Mainline implementation.
> • A PE without the Main Extension is also referred to as a Baseline implementation. A Baseline
> implementation has a subset of the instructions, registers, and features, of a Mainline
> implementation.
> • Armv7-M compatibility requires the Main Extension.
> • Armv6-M compatibility is provided by all Armv8-M implementations.

Trying to figure out under what case would we have something its ARMV6 only (and not baseline) and something that is ARMV7 only (and not mainline)?

ioannisg commented:
@galak I understand "backwards-compatibility" for Armv8-M Baseline in the following way: we shall be able to generate some code using Armv8-M_BASELINE and execute on a CPU that supports Armv6-M architecture.
In other words, Armv8-m_BASELINE shall have all elements (e.g. instruction set, rules, registers, behavior etc.) that are also present in Armv6-M.

The opposite does not need to hold **in theory**: Armv6-M could have additional elements. 

But I thought not to assume 1-1 correspondence between the two, because ARM does not claim it.

galak commented:
> @galak s/comments/commits ?

Yes, commits :)

GAnthony commented on arch/arm/soc/ti_simplelink/cc32xx/soc.h:
```diff
@@ -21,11 +21,11 @@ enum {
 	NonMaskableInt_IRQn           = -14,
 	HardFault_IRQn                = -13,
-#if defined(CONFIG_ARMV7_M)
+#if defined(CONFIG_ARMV7_M) || defined(CONFIG_ARMV8_M_MAINLINE)
```
I'm not seeing why this addition is needed, as the cc32xx soc is an ARMV7_M.

Under what conditions would CONFIG_ARMV8_M_MAINLINE be defined for the cc32xx?

It seems programs including this cc32xx soc.h file would have SOC_SERIES_CC32XX defined, which through arch/arm/sock/ti_simplelink/cc32xx/Kconfig.series, selects CPU_CORTEX_M4, which via arch/arm/core/cortex_m/Kconfig then selects ARMV7_M.

galak replied:
Any idea why the ifdef is there? Or any issue if we remove it?

GAnthony replied:
This is taken originally from cmsis.h.
But, by the reasoning above (i.e., that CONFIG_ARMV7_M must be set anyways), there should be no issue to remove it.

So, I just removed it from soc.h, and sanitycheck passed.  Result: PR#6009.

So, @ioannisg, I think you can remove the change to ti_simplelink/cc32xx/soc.h from your commit.

ioannisg replied:
Removed.

GAnthony reviewed (commented):
One question:

ioannisg committed:
arch: arm: common Armv8-M support

This PR includes the required changes in order to support
conditional compilation for Armv8-M architecture. Two
variants of the Armv8-M architecture are defined:
- the Armv8-M Baseline (backwards compatible with ARMv6-M),
- the Armv8-M Mainline (backwards compatible with ARMv7-M).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: define Cortex-M23, Cortex-M33 CPUs

This commit defines the Kconfig options for
ARM Cortex-M23 and Cortex-M33 CPUs. It also
udpates the generic memory map for M23 and M33
implementations.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

galak commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -65,16 +65,26 @@ config CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
+	  is also referred to as a Baseline Implementation. A
+	  Baseline implementation has a subset of the instructions,
+	  registers, and features, of a Mainline implementation.
+	  - Armv6-M compatibility is provided by all Armv8-M
```
Can we use ARMv6 and ARMv8, I know it differs from Arm's new branding, but its so much easier to read (same comment in other places).

ioannisg replied:
@galak  Fixed in the new Kconfig definitions.

galak commented on dts/arm/armv8-m.dtsi:
```diff
@@ -0,0 +1,25 @@
+
+		nvic: interrupt-controller@e000e100  {
+			compatible = "arm,v8m-nvic";
+			reg = <0xe000e100 0xc00>;
```
Need a binding to match arm,v8m-nvic (dts/bindings/interrupt-controller/arm,v8m-nvic.yaml)

ioannisg replied:
Fixed, thanks! Added it to the second patch (namely, for Cortex-M33, consistent with the other .yaml bindings for CM0+ and CM4, respectively)

galak commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -176,8 +211,8 @@ config XIP
 
-menu "ARM Cortex-M0/M0+/M3/M4/M7 options"
-	depends on ARMV6_M || ARMV7_M
+menu "ARM Cortex-M0/M0+/M3/M4/M7/M23/M33 options"
```
Nit: the M23/M33 should be in the second patch.

ioannisg replied:
Sure

galak reviewed (changes requested):
Looking good, some minor style things and missing binding file for v8m-nvic.

galak commented on dts/bindings/interrupt-controller/arm,v8m-nvic.yaml:
```diff
@@ -0,0 +1,32 @@
+---
+title: ARM Cortex M33 NVIC Interrupt Controller
```
We shouldn't use Cortex M33 in the title/description (this is a mistake in the v6m & v7m yaml's as well).  Let just use some thing like V8M in this file and move this change to the first commit.

ioannisg replied:
@galak I would like also to fix the titles for the other .yaml files, to have some level of consistency.

ioannisg replied:
Added a third commit with the style updates on the v6m and v7m bindings.

galak reviewed (changes requested):
some minor cleanups on the yaml

ioannisg committed:
arch: arm: minor style fixes in dts binding .yaml files

This commit updates the title and the description of the
ARMv6-M, ARMv7-M NVIC dts bindings, so that both the title
and the description reference the respective ARM architecture
and not the Cortex M cores.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
